src/VirtoCommerce.PricingModule.Data/Services/PricingSearchServiceImpl.cs

```diff
@@ -59,9 +59,22 @@ public virtual Task<PriceSearchResult> SearchPricesAsync(PricesSearchCriteria cr
 
                     if (criteria.GroupByProducts)
                     {
-                        var groupedQuery = query.GroupBy(x => x.ProductId).OrderBy(x => 1);
+                        var groupedQuery = query.Select(x => x.ProductId).OrderBy(x => x).Distinct();
                         result.TotalCount = await groupedQuery.CountAsync();
-                        query = groupedQuery.Skip(criteria.Skip).Take(criteria.Take).SelectMany(x => x);
+
+                        if (criteria.Take > 0)
+                        {
+                            if (criteria.ProductIds.IsNullOrEmpty())
+                            {
+                                criteria.ProductIds = await groupedQuery.OrderBy(x => x).Skip(criteria.Skip).Take(criteria.Take).ToArrayAsync();
```

Comment:
we don't need to change `criteria.ProductIds`  because it already handled in `BuildQueryAsync`
thus, you can replace the if/else statement with only single line
```C#
query = query.Where(x => groupedQueryids.Contains(x.ProductId));
```